framework/frequency_manager.hpp

```diff
@@ -15,15 +16,22 @@ class FrequencyManager
 private:
 
 #ifdef __linux__
-    int max_frequency_supported = 0;
-    int min_frequency_supported = 0;
+    // core-frequency variables
+    int max_core_frequency_supported = 0;
+    int min_core_frequency_supported = 0;
     std::vector<std::string> per_cpu_initial_scaling_governor;
     std::vector<std::string> per_cpu_initial_scaling_setspeed;
     int current_set_frequency = 0;
-    std::vector<int> frequency_levels;
-    int frequency_level_idx = 0;
+    std::vector<int> core_frequency_levels;
+    int core_frequency_level_idx = 0;
     static constexpr int total_frequency_levels = 9;
```

Comment:
If core and uncore are to share total number of frequency levels, I don't think we need to maintain separate frequency level indices for them. If there's the same number of levels for core and uncore, and frequencies changes in lockstep with every fracture, then `uncore_frequency_level_idx` and `core_frequency_level_idx` would always have the same value.

However, I think `total_frequency_levels` should also be split between core and uncore to give us better control over uncore frequencies. Uncore typically has smaller dynamic range of frequency control, so it probably won't need 9 different levels.

Reply:
You are right on the first part. But if we decide to implement second suggestion i.e., having different `total_frequency_levels` for core and uncore, wouldn't this require separate `frequency_level_idx` since the frequency_idx has to wrap over once all the levels have been covered?

Reply:
Yes, it's either-or, can't have both. If you choose to keep the same number of frequency levels, then code can be simplified. I would, however, opt for having different `total_frequency_levels` for core and uncore.

Reply:
Can you suggest a number that's ideal here? I am not an uncore expert. I will try to make it work on any number.

Reply:
I'd suggest `min(16, max_core_frequency_supported - min_core_frequency_supported)` for cores, `min(8, max_uncore_frequency_supported - min_uncore_frequency_supported)` (adjusted for units in which the max/min values are represented).

Reply:
suppose the `max_core_frequency_supported` = 3.6Ghz and `min_core_frequency_supported` is 1.2Ghz, min(16, 3.6-1.2) = min(16, 2.4) = 2.4 ? 

Similarly `max_uncore_frequency_supported` = 2.8Ghz and `min_uncore_frequency_supported` = 1.2Ghz, min(8, 2.8-1.2) = min(8, 1.8) = 1.6? 

Is my math right here? What am i missing?

Reply:
Sorry, my bad for using frequency multipliers. If you assume base clock is 100MHz (which it is), then min/max frequencies can be expressed as multiplies of 100MHz. For values you provided, this would translate to:
* `min(16, 36 - 12) = min(16, 24) = 16`
* `min(8, 28 - 12) = min(8, 18) = 8`
resulting in 16 frequency levels for cores and 8 for uncore. With number of levels we can calculate actual frequency as `f_min + N * f_step`, where N is in range(0, 16), or (0, 8) for uncore, and `f_step = (f_max - f_min) / N * 100MHz`.

Reply:
Done. Can you check and approve? I'll squash all the commits to first 2 commits and merge.